File: Flood-Fill.java
// Time Complexity : O(m * n)  [iterating over each pixel]
// Space Complexity : O(m * n)  [size of call stack calling dfs]
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Iterate over each item in the grid
// Perform dfs on this item
// In dfs if color is orignal color then change it to newColor
class FllodFillSolution {
    int newColorToApply;
    int orignalColor;
    int[][] dirs = { { -1, 0 }, { 0, 1 }, { 1, 0 }, { 0, -1 } };
    int m, n;

    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        // null check
        if (image == null || image.length == 0 || image[sr][sc] == newColor)
            return image;

        newColorToApply = newColor;
        orignalColor = image[sr][sc];

        m = image.length;
        n = image[0].length;

        dfs(image, sr, sc);

        return image;
    }

    private void dfs(int[][] image, int r, int c) {
        // base
        if (r < 0 || r == m || c < 0 || c == n || image[r][c] != orignalColor)
            return;

        // logic
        image[r][c] = newColorToApply;
        for (int[] dir : dirs) {
            dfs(image, r + dir[0], c + dir[1]);
        }
    }
}

File: flood-fill.js
// Time Complexity : O(M*N) for BFS and DFS
// Space Complexity : O(M*N) for BFS and DFS
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/**
 * @param {number[][]} image
 * @param {number} sr
 * @param {number} sc
 * @param {number} newColor
 * @return {number[][]}
 */
// DFS (iterative)
// var floodFill = function(image, sr, sc, newColor) {
//     if (image == null || image.length == 0) return image;
//     if (image[sr][sc] == newColor) return image;
    
//     let original = image[sr][sc],
//         stack = [[sr, sc]],
//         dirs = [[1, 0], [-1, 0], [0, 1], [0, -1]],
//         m = image.length,
//         n = image[0].length;
//     image[sr][sc] = newColor;
    
//     while (stack.length) {
//         let curr = stack.pop();
//         for (const dir of dirs) {
//             let dr = dir[0] + curr[0],
//                 dc = dir[1] + curr[1];
            
//             if (dr >= 0 && dr < m && dc >= 0 && dc < n && image[dr][dc] == original) {
//                 image[dr][dc] = newColor;
//                 stack.push([dr, dc]);
//             } 
//         }
//     }
//     return image;
// };

// DFS (recursive)
let dirs, m, n, original;
var floodFill = function(image, sr, sc, newColor) {
    if (image == null || image.length == 0 || image[sr][sc] == newColor) return image;
    
    original = image[sr][sc];
    dirs = [[1, 0], [-1, 0], [0, 1], [0, -1]],
    m = image.length,
    n = image[0].length;
    dfs(image, sr, sc, newColor);
    return image;
};

var dfs = function(image, r, c, newColor) {
    // Base Case
    if (image[r][c] != original) return;
    
    // Logic
    image[r][c] = newColor;
    for (const dir of dirs) {
        let dr = dir[0] + r;
        let dc = dir[1] + c;
        if (dr >= 0 && dr < m && dc >= 0 && dc < n) dfs(image, dr, dc, newColor);
    }
};

// BFS
// var floodFill = function(image, sr, sc, newColor) {
//     if (image == null || image.length == 0 || image[sr][sc] == newColor) return image;
    
//     let original = image[sr][sc],
//         queue = [[sr, sc]],
//         dirs = [[1, 0], [-1, 0], [0, 1], [0, -1]],
//         m = image.length,
//         n = image[0].length;;
//     image[sr][sc] = newColor;
    
//     while (queue.length) {
//         let curr = queue.shift();
        
//         for (const dir of dirs) {
//             let dr = dir[0] + curr[0],
//                 dc = dir[1] + curr[1];
            
//             if (dr >= 0 && dr < m && dc >= 0 && dc < n && image[dr][dc] == original) {
//                 image[dr][dc] = newColor;
//                 queue.push([dr, dc]);
//             }
//         }
//     }
//     return image;
// };
